README
============================================================================

Prometheus QoS - steal fire from your ISP !
"fair-per-IP" quality of service utility
requires Linux kernel with HTB qdisc enabled

Copyright(C) 2005-2008 Michael Polak (xChaos) original source code
Copyright(C) 2007-2008 Martin Svoboda (Ludva) iptables-restore, CLASSIFY

Credit: CZFree.Net+Netdave (idea), Aquarius (.rpm), Gandalf (.deb)
...and: Martin Devera (.cz) for his HTB qdisc (of course) 
...and: Jakub Walczak (.pl) for providing feedback and patches
...and: Ing. Jiri Engelthaler (.cz) for bugfixes and Asus WL500 port
...and: Dial Telecom (our slightly expensive ISP) for chance to test it

Feedback: xchaos(at)arachne.cz
Homepage: http://gpl.arachne.cz
SVN tree: https://dev.arachne.cz/svn/prometheus
============================================================================

QoS (or Quality-of-service) is IPv4 traffic shaper replacement for Internet
Service Providers (ISP). Dump your vintage hard-wired routers/shapers
(C|sco, etc.) in favour of powerful open source and free solution !

Prometheus QoS generates multiple nested HTB tc classes with various rate 
and ceil values, and implements optional daily traffic quotas and data
transfer statistics (as HTML). It is compatible with NAT, both asymetrical
and symetrical, yet still provides good two-way shaping and prioritizing,
both upload and download. Prometheus QoS allows both "hard shaping" 
(reducing HTB ceil value for aggressive downloaders) and "soft shaping" 
(keeping HTB ceil, but reducing HTB prio, probably optimal solution for 
normal users).

Prometheus iGW was written in C<<1, which means it compiles simply with 
GNU C Compiler, and doesn't require any external liberaries (except libc)
and huge interpreter packages (like Perl or Java) to run. However, it 
depends on HTB algorithm hardcoded in Linux kernel. It is currently being 
tested in real-world enviroment to provide QoS services on 30 Mbps internet
gateway and proxy being used by 2000+ PCs connected to gateway using
CZFree.Net broadband community network.

Advantages over more straightforward traffic control scripts include
HTB fine tuning features (rate and ceil magic), data transfer statistics,
optional data transfer quotas, full NAT (both symetric and one way)
compatibility and optinal sharing of bandwith by IPs in completely 
different subnets.

Performance and scaling - current release: 

we run Prometheus QoS on Celeron 2.8 Ghz serving around 600 individual
traffic classes (fine tuning is using five user-defined prometheus.conf 
keyword) and another 2000 IPs sharing bandwith with certain other IPs 
("sharing-" keyword). Prometheus QoS is especially strong tool if you want 
IP's from different subnets to share the same traffic class.

With 30 Mbps (each way) total capacity of line, Cisco Catalyst 2950 on 
ISP side and up to cca 6000 packets per seconds, we were running into some 
problems with overall system load. We moved QoS from Athlon 1700 XP to 
Celeron 2.8 Ghz, and kept all SNAT related stuff (see optinal-tools directory)
on Athlon 1.7 Ghz, which alowed for peak throughput up to 10000 pps.

Performance fine tuning - history:

With kernel version 2.4.20 and release 0.2 we started to experience problems
at cca 1500 packets/sec. However, with new iptables indexing feature
implemented in 0.3 release, system load seems to be approximately 
10 times lower. Same HW was later shaping 2000 packets/sec without problems,
and it looked like comparable relatively low-end system should be able to do
traffic shaping for at least 10000 packets/sec (well, if HotSaNIC was turned
off, of course <g>). With 0.6 release and dynamicaly calculated iptables
indexing scheme we made it up to 6000 packets/sec, and then we ran into some
performance-related problems, which may be related to the fact we are doing
SNAT of 1000+ individual IP addresses on the same machine which is doing 
also the QoS: something on the way seems to be limited to 34 Mbps HD
(half-duplex, sum of upload and download) no matter what we try. Our ISP 
claims the fault is not on his side, so our next step will be to separate
traffic shaping and massive SNAT (IP masquerading) and assign separate
PC-based router to do each task.

Maximum performance observed with prometheus 0.6 and hashtable optimization
of tables with individual SNAT targets was up to 9000 packets/sec at cca 40
Mbps half-duplex (more then 20 Mbps fyull-duplex). However, this required
massive optimization, including 
echo -n 65000 > /proc/sys/net/ipv4/ip_conntrack_max 
and
echo -n 21600 > /proc/sys/net/ipv4/netfilter/ip_conntrack_tcp_timeout_established
and disabling of most userspace applications (like eg. hotsanic). At the
same time, router machine and system was accumulating wide set of various
performance related problems, which required us to reboot it at least
mohtly.

Note: Some time ago it seemed that maximum index of tc classes was restricted
to 10000 - but I haven't checked this again for quite a while.
Another note: All the echo stuff in previous paragraph can be also achieved by
adding following lines to /etc/sysctl.conf which is much cleaner way to do it:
sys.net.ipv4.ip_conntrack_max=65000
sys.net.ipv4.netfilter.ip_conntrack_tcp_timeout_established=21600

Future plans include also setting of individual daily limits on maximum
pps (packets per second) rates allocated to individual IP addresses (this
may be needed partly because of problems mentioned above).

============================================================================

Prometheus QoS is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as 
published by the Free Software Foundation; either version 2.1 of 
the License, or (at your option) any later version.

Prometheus QoS is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License 
along with Prometheus QoS source code; if not, write to 
Michael Polak, Svojsikova 7, 169 00 Praha 6 Czech Republic


CHANGELOG
version 0.8.0 2008-11-25

- HTB priorities behaviour fixed (magic-priorities keyword is now obsolete)
- prometheus now requires -r ("run") switch to start shaping, default is no action
- C<<1 dialect upgraded to 0.6.2 (for_each() { } instead of every() { }, etc.)
- several minor changesets at https://dev.arachne.cz/svn/prometheus/timeline

version 0.8.0 2008-09-05

- integrated monthly and yearly summary of traffic logs into prometheus script
- bugfix in changeset 75 - 91

version 0.7.8 2008-02-02 (xChaos)

- improved Enterprise Research and Planning (ERP) section
- added informational messages before running iptables-restore
- fixed behaviour of "...-magic-relative-prio 0" + "...-fixed-fixed-prio 0"
- "filter-type classify" prometheus.conf keyword tested in production enviroment
- log file /var/log/prometheus renamed to /var/log/prometheuslog
- credit file /var/run/prometheus.credit moved to  /var/lib/misc

version 0.7.7 2008-01-??

- testing version, 0.8 will be stable release
- faster startup times using iptables-restore (credit: Martin "Ludva" Svoboda)
- compiles and runs correctly on both 32bit and 64bit platforms

version 0.7 2007-01-29

- There were good reasons to rename iptables hashing to iptables indexing.
  Hashing is different algorithm, which offers constant complexity 
  independent on number of hashed entries and high memory requirements, while 
  iptables indexing used in Prometheus offers log8(N) complexity and relatively
  low memory requirements compared to hashing. 
- Multiple keywords can be defined and used in /etc/hosts-like nodelist,
  which allows definition of various data limits for different users or
  definition of both symetrical and asymetrical virtual lines in same config.
- Following prometheus.conf variables were made keyword-dependent:
  KEYWORD-magic-relative-limit, KEYWORD-magic-relative-prio,
  KEYWORD-magic-fixed-limit, KEYWORD-magic-fixed-prio, KEYWORD-,
- Following new prometheus.conf keyword-dependent variables were added:
  KEYWORD-html-color
- Following prometheus.conf variables were removed as obsolete and confusing:
  htb-ceil-bonus-divide, htb-ceil-divide

version 0.6 2006-06-29

- Hyper-optimized iptables hashing (automatic, but compile-time configurable)
- Option hashtable-treshold was made now obsolete
- No "Proxy" column in statistics when qos-proxy-enable is set to 0
- iptables verbose output parsing was rewritten
- Unfinished implementation of some pps limiting and soft-restart related code
- Maybe I have changed something else too, but it was long time since last release

version 0.5.3 2005-08-15

- Predictable error messages generated by tc and iptables were
  redirected to /dev/null
- magic-fixed-prio and magic-relative-prio can be used even with
  magic-fixed-limit and magic-relative-limit set to 0
- Minor bugs in data initializtion were fixed, so program should 
  compile and run also on exotic non-X86 plaforms like BusyBox Linux

version 0.5.2 2005-06-19

- Fixed long integer traffic counter overflow at 4 gigabytes (up or down)
- Fixed incorrect inclusion of http proxy traffic in direct traffic
- Fixed incorrectly displayed parameters of asymetrical upload (Jakub Walczak)
- Dry run (-d switch) output beautification
- Added check if minimum data transfer value is greater then 0 kbps
- magic-priorities added, allows better soft-shaping wihtout reducing HTB ceil

version 0.5.1 2005-03-22

- Fixed serious bug in 0.5 release

version 0.5 2005-03-21

- Better compatibility with up-to-date gcc and new HTB in 2.6.x kernel
- Enterprise Research and Planning (ERP) analysis implemented
- Credit file support implemented (for use with magic-relative-limit)
- magic-relative-limit-4prio now allows softer FUP using HTB prio
- digital-divide-weirdness-ratio added, only tested value is "2"
- qos-leaf-discipline and other new prometheus.conf keywords added

version 0.4 2004-11-02

- HTTP proxy traffic accounting bug fixed (occured with hashed iptables)
- free bandwith traffic shaping (allowed unlimited upload with hashing)
- maximal guaranted speed increased to 2048 kbps (just tc namespace problem)
  (Read: it seems that maximum index of tc classes is restricted to 10000)
- qos-proxy-ip can be inside qos-free-zone even if qos-proxy-enable 
  (Read: Single machine can run NAT/firewall, HTTP proxy and Prometheus QoS)
- command line argument -n (no delay) overrides qos-free-delay keyword
  (Read: Separate configuration file for first run and daily run is not needed)
- traffic from unregistered IPs is monitored even with hashing enabled

version 0.3 2004-10-13

- massive performance improvement thanks to hashing of iptables
- hashtables for /24 subnets are forced if there are 256+ traffic classes
- hashtables can be forced using new option hashtable-treshold)
- qos-free-zone is now optional (use qos-free-zone 0)
- minor bug fix: -h command line arugment now overrides config file

version 0.2 2004-09-20

- daily logging of data transfered in individual traffic classes
- preview feature (create HTML hall of fame without reseting counters) 
- delay feature (useful for creating QoS-free period without reseting counters)
- <A NAME=""> anchors with hostnames of individual traffic classes
- tested succesfuly on our upgraded 10Mbps-full duplex, <1000 packets/sec

version 0.1 

- initial release
- tested succesfuly on our 3Mbps-full duplex line


cll1-0.6.2.h

/* C<<1 header file v0.6.2 - style sheet for ANSI C  */
/* Please pronounce as "cee-shift-left-by-one" :)  */

/* Copyright (G) 2004-2008 Michael xChaos Polak, x(at)n.cz

   The C<<1 header file is free software; you can redistribute it and/or
   modify it under the terms of the GNU Lesser General Public
   License as published by the Free Software Foundation; either
   version 2.1 of the License, or (at your option) any later version.

   The C<<1 header file is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with the GNU C Library; if not, write to 
   Michael Polak, Svojsikova 7, 169 00 Praha 6 Czech Republic */


#ifndef __CLL1H__
#define __CLL1H__

#include <stdio.h>
#include <string.h>
#include <strings.h>
#include <malloc.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#include <ctype.h>

/* Section For Dummies part 1, updated 2004-05-07 by xCh. */

#define not !
#define TRUE 1
#define FALSE 0
#define loop while(1)
#define iterate(VAR,FROM,TO) for(VAR=FROM; VAR <= TO; VAR++)
#define repeat(N) iterate(_i,1,N)

/* Dynamic list macros & sequences, updated 2003-05-29 by xCh. */

#define list(T) struct T *_next 
#define create(A,T) (A=(struct T *)malloc(sizeof(struct T)),A->_next=NULL)
#define push(A,B) { if(A && A!=B) A->_next=B; B=A; }
#define append(A,B) { if(B) { void *N=A; A->_next=NULL; for_selected(A,B,!A->_next) {A->_next=N; break;}} else push(A,B); }
#define remove(A,B,C) { void **_D=NULL; for_selected(A,B,C) { if(_D)*_D=A->_next; else B=A->_next; free(A); } else _D=(void *)&(A->_next); }
#define drop(A,B) { for( A=B; A ; B=A, A=A->_next, free(B)); B=NULL; }

/* Dynamic list iterations and sequences, updated 2003-05-29 by xCh. */

#define for_each(A,B) for( A=B; A; A=A->_next)
#define for_selected(A,B,C) for_each(A,B) if(C)
#define if_exists(A,B,C) for_selected(A,B,C) break; if(A)

/* EXP macros for Dummysort sequences, updated 2003-05-29 by xCh. */

#define order_by(K1,K2) (K1>K2)
#define desc_order_by(K1,K2) (K1<K2)
#define sort_by(K1,K2) (strcasecmp(K1, K2)>0)
#define desc_sort_by(K1,K2) (strcasecmp(K1, K2)<0)
#define ascii_by(K1,K2) (strcmp(K1, K2)>0)
#define desc_ascii_by(K1,K2) (strcmp(K1, K2)<0)

/* Dummysort sequences, updated 2003-05-29 by xCh. */

#define insert(A,B,EXP,K) { if(B) { void **_L=NULL, *H=B; for_selected(B,H,EXP(B->K,A->K)) { if(_L) {*_L=A; A->_next=B; } else push(A,H); break; } else _L=(void *)&(B->_next); if(!B)*_L=A; B=H; } else push(A,B); }
#define sort(A,B,EXP,K) { void *_C; A=B; B=NULL; do { _C=A->_next; A->_next=NULL; insert(A,B,EXP,K); A=_C; } while(_C); }

/* String macros & sequences, updated 2004-04-19 by xCh. */

#define eq(A,B) !strcmp(A,B)
#define strcmpi(A,B) strcasecmp(A,B)
#define strlwr(A) {char *_S=A; while(_&&*_S){*_S=tolower(*_S);_S++;}}
#define strupr(A) {char *_S=A; while(_&&*_S){*_S=toupper(*_S);_S++;}}
#define string(S,L) (S=(char *)malloc(L),*S=0)
#define duplicate(A,B) if(A) { string(B,strlen(A)+1); strcpy(B,A); }
#define concatenate(A,B,C) if (A && B) { string(C,strlen(A)+strlen(B)+1); strcpy(C,A); strcat(C,B); }
#define suffix(A,B,C) (((A=strrchr(B,C))&&!(*(A++)=0))||(A=B))
#define prefix(A,B,C) ((A=B)&&(((B=strchr(B,C))&&!(*(B++)=0))||(B=A)))
#define gotoalpha(CHAR) if(CHAR)while(*CHAR && !isalpha(*CHAR))CHAR++
#define goto_alpha(CHAR) if(CHAR)while(*CHAR && !isalpha(*CHAR) && *CHAR!='_')CHAR++
#define gotoalnum(CHAR) if(CHAR)while(*CHAR && !isalnum(*CHAR))CHAR++
#define goto_alnum(CHAR) if(CHAR)while(*CHAR && !isalnum(*CHAR) && *CHAR!='_')CHAR++
#define skipalpha(CHAR) if(CHAR)while(*CHAR && isalpha(*CHAR))CHAR++
#define skip_alpha(CHAR) if(CHAR)while(*CHAR && (isalpha(*CHAR) || *CHAR=='_'))CHAR++
#define skipalnum(CHAR) if(CHAR)while(*CHAR && isalnum(*CHAR))CHAR++
#define skip_alnum(CHAR) if(CHAR)while(*CHAR && (isalnum(*CHAR) || *CHAR=='_'))CHAR++
#define skipspaces(CHAR) if(CHAR)while(*CHAR==' ')CHAR++
#define cutspaces(CHAR) if(CHAR){int _L=strlen(CHAR); while(--_L>0 && CHAR[_L]==' ')CHAR[_L]=0;}
#define gotochr(CHAR,C) if(CHAR)while(*CHAR && *CHAR!=C)CHAR++
#define tr(CHAR,B,C) {char *_S=CHAR; while(*_S){ if(*_S==B)*_S=C; _S++; }}
#define strswitch(CHAR) {char *_K=CHAR; FILE *_F=NULL; {{ 
#define stroption(STR) if(eq(STR,_K))
#define match(KEY,VAL) {char *_K=KEY, *_V=VAL; FILE *_F=NULL; {{ 
#define assign(STR,SETVAR) stroption(STR) SETVAR=_V

/* Section For Dummies part 2, updated 2004-05-07 by xCh. */

#define program int _I; int main(int argc, char **argv)
#define arguments if(argc>1) for(_I=1;_I<argc;_I++) 
#define argument(A) if(eq(argv[_I],A))
#define thisargument(S) (S=argv[_I])
#define nextargument(S) if(_I+1<argc && (S=argv[++_I]))

/* I/O iterations, updated 2004-04-19 by xCh. */

#define fparse(S,L,F) for(fgets(S,L,F);*S && !feof(F);fgets(S,L,F))
#define input(S,L) fparse(S,L,stdin)
#define fstring(S,F) { int _C=0,_L=0; fpos_t _P; fgetpos(F,&_P); while(_C!='\n' && !feof(F)){ _C=fgetc(F); _L++; } string(S,_L); fsetpos(F,&_P);fgets(S,_L,F);fgetc(F);}
#define parses(S,F) {FILE *_F=fopen(F,"r"); if(_F) { while(!feof(_F)) { fstring(S,_F);  
#define parse(F) {char *_; FILE *_F=fopen(F,"r"); if(_F) { while(!feof(_F)) { fstring(_,_F);  
#define fail }} else {{
#define done }} if(_F)fclose(_F);}
#define option(STR,SETVAR) if(_){char *_K,*_V,*_O,*_Q; duplicate(_,_Q); _O=_Q; tr(_O,'\t',' '); prefix(_K,_O,' '); if(eq(STR,_K)) {skipspaces(_O); prefix(_V,_O,'#'); cutspaces(_V); SETVAR=_V; _=NULL;} else free(_Q);}
#define ioption(STR,SETVAR) if(_){char *_K,*_V,*_O,*_Q; duplicate(_,_Q); _O=_Q; tr(_O,'\t',' '); prefix(_K,_O,' '); if(eq(STR,_K)) {skipspaces(_O); prefix(_V,_O,'#'); cutspaces(_V); SETVAR=atoi(_V); _=NULL;} free(_Q);}
#define loption(STR,SETVAR) if(_){char *_K,*_V,*_O,*_Q; duplicate(_,_Q); _O=_Q; tr(_O,'\t',' '); prefix(_K,_O,' '); if(eq(STR,_K)) {skipspaces(_O); prefix(_V,_O,'#');  cutspaces(_V); SETVAR=atol(_V); _=NULL;} free(_Q);}
#define lloption(STR,SETVAR) if(_){char *_K,*_V,*_O,*_Q; duplicate(_,_Q); _O=_Q; tr(_O,'\t',' '); prefix(_K,_O,' '); if(eq(STR,_K)) {skipspaces(_O); prefix(_V,_O,'#');  cutspaces(_V); SETVAR=atoll(_V); _=NULL;} free(_Q);}

/* Dynamic list advanced I/O, updated 2003-05-30 by xCh. */

#define load(A,B,F,T,K) {char *_S; parses(_S,F) { create(A,T); A->K=_S; A->_eoln=TRUE; append(A,B);} done; A->_eoln=FALSE;}
#define save(A,B,F,K) {FILE *_F=fopen(F,"w"); if(_F) { for_each(A,B) {fputs(A->K,_F); if(A->_eoln) fputc('\n',_F);} fclose(_F);}}
     
/* I/O sequences, updated 2003-05-29 by xCh. */

#define nullreopen(F) F=freopen("/dev/null","r",F)
#define stdinredir(CMD) {int _r[2];pipe(_r);if(fork()==0){dup2(_r[1],1);close(_r[0]);CMD;exit(0);}nullreopen(stdin);dup2(_r[0],0);close(_r[1]);}
#define shell(CMD) stdinredir(system(CMD))
#define paste(STR) stdinredir(fputs(STR,stdout))

/* String iterations, updated 2003-06-19 by xCh. */

#define split(A,B,C) for(prefix(A,B,C);A;(A!=B)&&prefix(A,B,C)||(A=NULL))
#define valid_split(A,B,C) split(A,B,C) if(*A)
#define columns(A,B,C,V) for(V=0,prefix(A,B,C);A;((A!=B)&&prefix(A,B,C)||(A=NULL)),V++)
#define valid_columns(A,B,C,V) for(V=0,prefix(A,B,C);A;((A!=B)&&prefix(A,B,C))||(A=NULL)) if(*A&&++V)
#define column(A,B,C,V) { int _V; columns(A,B,C,_V) if(_V==V) break; }
#define valid_column(A,B,C,V) { int _V; valid_columns(A,B,C,_V) if(_V==V) break; }

/* Useful structures, updated 2003-05-29 by xCh. */

#define hashtable(TYPE,NAME,VALUE) struct TYPE { char *NAME; char *VALUE; list(TYPE); }
#define textfile(TYPE,LINE) struct TYPE { char *LINE; char _eoln; list(TYPE); }
#define date(S) { time_t _T; _T=time(NULL); duplicate(ctime(&_T),S); }

#endif
